File: ios/RBDarts/Features/StandaloneGame/StandaloneGameService.swift
import Foundation

public actor StandaloneGameService {
    private let repository: RBDartsRepository
    private let sessionStore: ActiveScoreSessionStoring

    public init(
        repository: RBDartsRepository = InMemoryRBDartsRepository(),
        sessionStore: ActiveScoreSessionStoring = FileActiveScoreSessionStore()
    ) {
        self.repository = repository
        self.sessionStore = sessionStore
    }

    public func createGame(
        name: String,
        createdByUserId: String,
        participantNames: [String],
        regulationInnings: Int = ScoringRules.defaultRegulationInnings,
        extraInningsEnabled: Bool = true
    ) async throws -> ActiveScoreSession {
        let now = Date()
        let game = StandaloneGame(
            standaloneGameId: UUID().uuidString,
            createdByUserId: createdByUserId,
            gameName: name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "Standalone Game" : name,
            gameDate: now,
            participantType: .singles,
            regulationInnings: regulationInnings,
            extraInningsEnabled: extraInningsEnabled,
            status: .scoring,
            createdAt: now,
            completedAt: nil
        )
        let participants = participantNames
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .enumerated()
            .map { index, name in
                ParticipantScorecard(
                    participantId: "participant-\(index + 1)",
                    displayName: name.trimmingCharacters(in: .whitespacesAndNewlines)
                )
            }
        let safeParticipants = participants.isEmpty ? [
            ParticipantScorecard(participantId: "participant-1", displayName: "Player 1")
        ] : participants

        let session = ActiveScoreSession(
            sessionId: game.standaloneGameId,
            updatedAt: now,
            game: game,
            participants: safeParticipants,
            pendingSyncRequestIds: []
        )
        try sessionStore.save(session)
        try await repository.createStandaloneGame(game, participants: safeParticipants)
        return session
    }

    public func enterScore(
        _ score: Int,
        participantId: String,
        inningNumber: Int,
        in session: ActiveScoreSession
    ) async throws -> ActiveScoreSession {
        var updated = session
        guard let index = updated.participants.firstIndex(where: { $0.participantId == participantId }) else {
            return session
        }
        try ScoringRules.recordScore(
            score,
            inningNumber: inningNumber,
            for: &updated.participants[index]
        )
        updated.updatedAt = Date()
        try sessionStore.save(updated)
        return updated
    }

    public func undoScore(
        participantId: String,
        inningNumber: Int,
        in session: ActiveScoreSession
    ) throws -> ActiveScoreSession {
        var updated = session
        if let index = updated.participants.firstIndex(where: { $0.participantId == participantId }) {
            updated.participants[index].scoresByInning.removeValue(forKey: inningNumber)
            updated.updatedAt = Date()
            try sessionStore.save(updated)
        }
        return updated
    }

    public func complete(_ session: ActiveScoreSession) async throws -> GameSummary {
        var completed = session
        completed.game.status = .complete
        completed.game.completedAt = Date()
        let summary = ScoringRules.summarizeStandalone(
            participants: completed.participants,
            regulationInnings: completed.game.regulationInnings,
            extraInningsEnabled: completed.game.extraInningsEnabled
        )
        try sessionStore.save(completed)
        try await repository.saveStandaloneSummary(summary, for: completed.game.standaloneGameId)
        return summary
    }

    public func recoverMostRecentSession() throws -> ActiveScoreSession? {
        try sessionStore.loadMostRecent()
    }
}

File: android/app/src/main/java/com/rbdarts/feature/standalonegame/StandaloneGameService.kt
package com.rbdarts.feature.standalonegame

import com.rbdarts.core.data.RBDartsRepository
import com.rbdarts.core.domain.GameStatus
import com.rbdarts.core.domain.GameSummary
import com.rbdarts.core.domain.ParticipantScorecard
import com.rbdarts.core.domain.ParticipantType
import com.rbdarts.core.domain.ScoringRules
import com.rbdarts.core.domain.StandaloneGame
import java.time.Instant
import java.util.UUID

data class ActiveScoreSession(
    val sessionId: String,
    val updatedAt: Instant,
    val game: StandaloneGame,
    val participants: List<ParticipantScorecard>,
    val pendingSyncRequestIds: List<String> = emptyList()
)

interface ActiveScoreSessionStore {
    suspend fun save(session: ActiveScoreSession)
    suspend fun loadMostRecent(): ActiveScoreSession?
    suspend fun delete(sessionId: String)
}

class InMemoryActiveScoreSessionStore : ActiveScoreSessionStore {
    private var latest: ActiveScoreSession? = null

    override suspend fun save(session: ActiveScoreSession) {
        latest = session
    }

    override suspend fun loadMostRecent(): ActiveScoreSession? = latest

    override suspend fun delete(sessionId: String) {
        if (latest?.sessionId == sessionId) latest = null
    }
}

class StandaloneGameService(
    private val repository: RBDartsRepository,
    private val sessionStore: ActiveScoreSessionStore = InMemoryActiveScoreSessionStore()
) {
    suspend fun createGame(
        name: String,
        createdByUserId: String,
        participantNames: List<String>,
        regulationInnings: Int = ScoringRules.DefaultRegulationInnings,
        extraInningsEnabled: Boolean = true
    ): ActiveScoreSession {
        val now = Instant.now()
        val gameId = UUID.randomUUID().toString()
        val game = StandaloneGame(
            standaloneGameId = gameId,
            createdByUserId = createdByUserId,
            gameName = name.trim().ifBlank { "Standalone Game" },
            gameDate = now,
            participantType = ParticipantType.Singles,
            regulationInnings = regulationInnings,
            extraInningsEnabled = extraInningsEnabled,
            status = GameStatus.Scoring,
            createdAt = now,
            completedAt = null
        )
        val participants = participantNames
            .filter { it.isNotBlank() }
            .mapIndexed { index, participantName ->
                ParticipantScorecard(
                    participantId = "participant-${index + 1}",
                    displayName = participantName.trim()
                )
            }
            .ifEmpty {
                listOf(ParticipantScorecard(participantId = "participant-1", displayName = "Player 1"))
            }

        val session = ActiveScoreSession(
            sessionId = gameId,
            updatedAt = now,
            game = game,
            participants = participants
        )
        sessionStore.save(session)
        repository.createStandaloneGame(game, participants)
        return session
    }

    suspend fun enterScore(
        score: Int,
        participantId: String,
        inningNumber: Int,
        session: ActiveScoreSession
    ): ActiveScoreSession {
        val updatedParticipants = session.participants.map { participant ->
            if (participant.participantId == participantId) {
                ScoringRules.recordScore(participant, inningNumber, score)
            } else {
                participant
            }
        }
        val updated = session.copy(updatedAt = Instant.now(), participants = updatedParticipants)
        sessionStore.save(updated)
        return updated
    }

    suspend fun undoScore(
        participantId: String,
        inningNumber: Int,
        session: ActiveScoreSession
    ): ActiveScoreSession {
        val updatedParticipants = session.participants.map { participant ->
            if (participant.participantId == participantId) {
                participant.copy(scoresByInning = participant.scoresByInning - inningNumber)
            } else {
                participant
            }
        }
        val updated = session.copy(updatedAt = Instant.now(), participants = updatedParticipants)
        sessionStore.save(updated)
        return updated
    }

    suspend fun complete(session: ActiveScoreSession): GameSummary {
        val summary = ScoringRules.summarizeStandalone(
            participants = session.participants,
            regulationInnings = session.game.regulationInnings,
            extraInningsEnabled = session.game.extraInningsEnabled
        )
        repository.saveStandaloneSummary(summary, session.game.standaloneGameId)
        return summary
    }

    suspend fun recoverMostRecentSession(): ActiveScoreSession? = sessionStore.loadMostRecent()
}
